### DISCUSSION TITLE
Cannot create outline

### DISCUSSION THREAD

**Discussion by decoomanj** [James-Yu/LaTeX-Workshop#3399] [Q&A] [answered]
When I have a large LaTeX document, the outline cannot be generated. It works well for smaller files. When I chop up the large file in two smaller files, it works again. Is there a timeout/deadline for the generation of the outline? If this is the case, can this setting be changed?

I'm using the version of 7/25/2022. It worked up to last week.

**Comment by James-Yu** [comment]
You may need to provide a working example with randomly generated contents for debugging.

**Comment by decoomanj** [comment]
Sure, here you go. When the file is too large, the scanning just stops.

![image](https://user-images.githubusercontent.com/1961569/181189299-563b1223-56c3-40ff-a0d6-69a5a35c1b1d.png)

[test.tex.zip](https://github.com/James-Yu/LaTeX-Workshop/files/9196110/test.tex.zip)

**Comment by James-Yu** [comment]
<img width="903" alt="image" src="https://user-images.githubusercontent.com/4210342/181191358-bfdf877b-fb9a-45d8-b0e4-cecee4b321b0.png">

It shows, though slow.

**Reply by decoomanj** [reply]
Hm, what happens if you double the content again? On my machine it takes like more or less 5s until the operation terminates unsuccessfully.

**Reply by decoomanj** [reply]
Ah, by the way, I was referring to the outline panel of VSCode. But this is maybe not generated by your plugin, right?

**Comment by decoomanj** [comment]
Just found this in the logs:

[2022-07-27 10:52:16.289] [renderer1] [error] Cannot resolve tree item for element 0/0:0.2.2.2 \largesubsubsection: Error: Cannot resolve tree item for element 0/0:0.2.2.2 \largesubsubsection
	at /snap/code/102/usr/share/code/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:106:19334
	at processTicksAndRejections (node:internal/process/task_queues:96:5)

**Reply by James-Yu** [reply]
This is a known issue with the current structure provider. However, the bug might not be caused by the extension after my previous extensive inspection. I still have no idea on how this message is generated as it cannot be reproduced consistently.

**Comment by decoomanj** [comment]
Do you know which component is responsible to populate the outline-panel? (not the structure panel) 

I disabled your plugin, and then the outline is not filled. I suppose your plugin is supplying the data?

**Reply by decoomanj** [reply]
I suppose there is a regression in the speed of the outline/structure. Before the update, it was generated way faster. Now I have need 2 to 4 seconds for small files (3000 lines). Could it be that the document gets parsed too many times?

**Reply by decoomanj** [reply]
I've reverted to the previous release, and although it takes some time, the outline works again.

**Comment by James-Yu** [comment, marked as answer]
A PR #3400 is suggested to fix (to some extent) this issue.

The reason on the regression is that the extension shift to use AST parser to construct a document structure, instead of the previous regex-based approach. The new one has many benefits but one demerit: speed. Typically, LaTeX documents will be split into multiple small files if total lines exceeds several hundred, which is quite the best practice. In such cases, the new approach is notably superior.

Despite #3400 may solve the issue, I personally still recommend you to split huge documents into multiples. 3000-line LaTeX is (to myself) huge. Just my 2c.

**Reply by decoomanj** [reply]
Thanks for the quick response. I have already several tex-files, but the main one is around 25K lines. This actually worked quite well. If I need to split this file in smaller ones, then I have the disadvantage again that I need to search stuff in different files which makes writing more tedious to me. Building the AST seems the better approach, but when the creation of the outline is too slow, the user experience is not so great. Another question is why the creation of the AST fails on large files? Is there a timeout for this operation?

**Comment by rat10** [comment]
Which version did introduce the new AST-based code? Which version would I have to revert to to get the old regex-based approach?

**Comment by decoomanj** [comment]
Hi again, I was using an older version of the module to generate the outline. I have tested the new module, but unfortunately the option for the fastparser is not available. I need to deal with larger Latex files, and there is not a real way around it. Could the timeout for the AST calculation be extended/changed?
